core/src/main/java/org/mule/runtime/core/execution/ModuleFlowProcessingPhase.java

```diff
@@ -142,25 +144,26 @@ public void runPhase(final ModuleFlowProcessingPhaseTemplate template, final Mes
                                MESSAGE_RECEIVED);
             })
             .then(request -> from(template.routeEventAsync(request)))
-            .doOnSuccess(getSuccessConsumer(messageSource, templateEvent, messageProcessContext, phaseResultNotifier, template,
-                                            terminateConsumer))
-            .doOnError(MessagingException.class, me -> {
-              if (me.getCause() instanceof SourceErrorException
-                  && sourceResponseErrorTypeMatcher.match(((SourceErrorException) me.getCause()).getErrorType())) {
-                try {
-                  handleSourceError(exceptionHandler, errorConsumer, (SourceErrorException) me.getCause());
-                  onTerminate(terminateConsumer, left(me.getEvent()));
-                } catch (Exception e) {
-                  onTerminate(terminateConsumer, right(e));
-                }
-              } else {
-                try {
+            .doOnSuccess(successConsumer)
+            .onErrorResume(MessagingException.class, me -> {
+              try {
+                if (me.getCause() instanceof SourceErrorException
+                    && sourceResponseErrorTypeMatcher.match(((SourceErrorException) me.getCause()).getErrorType())) {
+                  handleSourceError(exceptionHandler, errorConsumer, me, terminateConsumer);
+                } else if (me.inErrorHandler()) {
```

Comment:
IMO rather than have a different code path for handling error handler exceptions like this, all exceptions from Flow should go via same (onError) route, and if required onError can do conditional behaviour based of if error was thrown by error handler or not by matching ErrorType.